src/mvba/abba/mod.rs

```diff
@@ -1,56 +1,567 @@
+                            (
+                                PreVoteValue::One,
+                                PreVoteJustification::HardJustification(sig.clone()),
+                            )
+                        } else if abstain_count == self.threshold() {
+                            // if all main-votes are abstain,
+                            let sig = match &main_votes.iter().last().unwrap().1.justification {
+                                MainVoteJustification::AbstainJustification(_, _) => {
+                                    let sig_share: HashMap<&NodeId, &SignatureShare> =
+                                        main_votes.iter().map(|(n, a)| (n, &a.sig_share)).collect();
+                                    self.pub_key_set.combine_signatures(sig_share)?
+                                }
+                                _ => {
+                                    return Err(Error::Generic(
+                                        "protocol violated, invalid main-vote justification"
+                                            .to_string(),
+                                    ))
+                                }
+                            };
+                            // soft pre-vote for 1
+                            // Coin value bias to 1 in weaker validity mode.
+                            (
+                                PreVoteValue::One,
+                                PreVoteJustification::SoftJustification(sig),
+                            )
+                        } else {
+                            return Err(Error::Generic(
+                                "protocol violated, no pre-vote majority".to_string(),
+                            ));
+                        };
+
+                        // Produce an S-signature share on the message `(ID, pre-vote, r, b)`
+                        let sign_bytes = self.pre_vote_bytes_to_sign(self.r, &value)?;
+                        let sig_share = self.sec_key_share.sign(sign_bytes);
+
+                        // Send to all parties the message `(ID, pre-vote, r, b, justification, signature share)`
+                        let pre_vote_message = Message {
+                            id: self.id.clone(),
+                            action: Action::PreVote(PreVoteAction {
+                                round: self.r,
+                                value,
+                                justification,
+                                sig_share,
+                            }),
+                        };
+
+                        self.broadcast(pre_vote_message)?;
+                    }
+                }
+            }
+
+            Action::PreVote(action) => {
+                if action.round != self.r {
+                    return Ok(());
+                }
+                // Let's check pre-votes:
+
+                // HECK FOR DECISION. Collect n −t valid and properly justified main-votes of round r .
```

Comment:
Shouldn't this check for decision be in the `MainVote` branch above?
```suggestion
                // CHECK FOR DECISION. Collect n −t valid and properly justified main-votes of round r .
```

The comment says that the check triggers once n-t *main-votes* are collected.

Reply:
@b00f I think you missed this one. I think we want to check for decision when we receive a Main vote, you have it in the `PreVote` case.